service/dap/server_test.go

```diff
@@ -1855,14 +1856,79 @@ func TestSetBreakpoint(t *testing.T) {
 
 					// Set at a line without a statement
 					client.SetBreakpointsRequest(fixture.Source, []int{1000})
-					expectSetBreakpointsResponse([]Breakpoint{{1000, "", false, "could not find statement"}}) // all cleared, none set
+					expectSetBreakpointsResponse(t, client, []Breakpoint{{1000, "", false, "could not find statement"}}) // all cleared, none set
 				},
 				// The program has an infinite loop, so we must kill it by disconnecting.
 				disconnect: true,
 			}})
 	})
 }
 
+func TestSetBreakpointWhileRunning(t *testing.T) {
+	runTest(t, "sleep", func(client *daptest.Client, fixture protest.Fixture) {
+		runDebugSessionWithBPs(t, client, "launch",
+			// Launch
+			func() {
+				client.LaunchRequest("exec", fixture.Path, !stopOnEntry)
+			},
+			// Set breakpoints
+			fixture.Source, []int{9},
+			[]onBreakpoint{{
+				execute: func() {
+					handleStop(t, client, 1, "main.f", 9)
+					client.NextRequest(1)
+					client.ExpectNextResponse(t)
+					client.ExpectStoppedEvent(t)
+					handleStop(t, client, 1, "main.f", 10)
+
+					// This will be a 10 second sleepy next, so we can reliably interrupt it
+					client.NextRequest(1)
+					client.ExpectNextResponse(t)
+
+					// Set breakpoint at line 11 right after sleep
+					client.SetBreakpointsRequest(fixture.Source, []int{11})
+					se := client.ExpectStoppedEvent(t)
+					if se.Body.Reason != "cancelled next" || !se.Body.AllThreadsStopped || se.Body.ThreadId != 0 && se.Body.ThreadId != 1 {
+						t.Errorf("\ngot  %#v\nwant Reason='cancelled next' AllThreadsStopped=true ThreadId=0/1", se)
+					}
+					expectSetBreakpointsResponse(t, client, []Breakpoint{{11, fixture.Source, true, ""}})
+					oe := client.ExpectOutputEvent(t)
+					warning := "WARNING: setting breakpoints halted execution and cancelled \"next\"\n"
+					if oe.Body.Category != "stderr" || oe.Body.Output != warning {
+						t.Errorf("\ngot  %#v\nwant Category='stderr' Output=%q", oe, warning)
+					}
+
+					// We must continue since next got cancelled
+					client.ContinueRequest(1)
+					client.ExpectContinueResponse(t)
+
+					// Continue stops at line 11 (verifies that setting breakpoint during next worked)
+					se = client.ExpectStoppedEvent(t)
+					if se.Body.Reason != "breakpoint" || !se.Body.AllThreadsStopped || se.Body.ThreadId != 1 {
+						t.Errorf("\ngot  %#v\nwant Reason='breakpoint' AllThreadsStopped=true ThreadId=1", se)
+					}
+					handleStop(t, client, 1, "main.f", 11)
+
+					// We coninue and loop once more
+					client.ContinueRequest(1)
+					client.ExpectContinueResponse(t)
+
+					// While its running, set breakpoint at line 10 and clear line 11
```

Comment:
maybe add something about how we expect not to hit line 11 again before this request gets processed for the same reason as above (10 second sleep)